### PR TITLE
feat(datasource:terraform): use official api for custom registries

### LINKED ISSUE #14602: Terraform modules discovery not working on gitlab.com terraform registry
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

renovate/renovate:31.97.3@sha256:58ab4f71169a955888c0a89c3c94e7adadfebee75679ebc42aff06f9fbc9e485

### Please select which platform you are using if self-hosting.

gitlab.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Hi,
I am trying to get renovate working with the terraform registry of gitlab.com. Private registries have been implemented with https://github.com/renovatebot/renovate/issues/6707 and gitlab.com should work as a private registry, too.

Unfortunately I get a http 404 on an API endpoint that is not described in the module registry spec and probably not implemented in gitlab.com:
https://www.terraform.io/internals/module-registry-protocol#list-available-versions-for-a-specific-module

I suspect  that the datasource for terraform-modules might not be quite up to spec and extract the versions from the endpoint `:namespace/:name/:provider/`

On registry.terraform.io this endpoint returns a json representation of the data displayed, for example:
https://registry.terraform.io/modules/terraform-google-modules/service-accounts/google
https://registry.terraform.io/v1/modules/terraform-google-modules/service-accounts/google returns a json that also contains the versions:
```
"versions":["0.1.0","0.1.1","1.0.0","2.0.0","2.0.1","2.0.2","3.0.0","3.0.1","4.0.0","4.0.1","4.0.2","4.0.3","4.1.0","4.1.1"]
```

I have limited knowledge in ts, but it looks like `/versions` is not called here, but the versions are extracted from the module root instead?
https://github.com/renovatebot/renovate/tree/main/lib/modules/datasource/terraform-module

### Relevant debug logs

<details><summary>Logs</summary>

```
{"name":"renovate","hostname":"runner-gkqlnmnx-project-11111111-concurrent-0kmfnl","pid":13,"level":20,"logContext":"hpPcfm_IMlLOzgfIYtpny","repository":"mygroup/infra/common/pocs/terraform-renovate-test","datasource":"terraform-module","lookupName":"gitlab.com/mygroup/mymodule/google","url":"https://gitlab.com/api/v4/packages/terraform/modules/v1/mygroup/mymodule/google","msg":"Datasource 404","time":"2022-03-10T16:34:24.687Z","v":0}
{"name":"renovate","hostname":"runner-gkqlnmnx-project-11111111-concurrent-0kmfnl","pid":13,"level":20,"logContext":"hpPcfm_IMlLOzgfIYtpny","repository":"mygroup/infra/common/pocs/terraform-renovate-test","dependency":"gitlab.com/mygroup/mymodule/google","packageFile":"terraform/common/main.tf","msg":"Failed to look up dependency gitlab.com/mygroup/mymodule/google","time":"2022-03-10T16:34:24.687Z","v":0}
```

</details>

### Have you created a minimal reproduction repository?

https://gitlab.com/renovate-issue-14602

### PR DESCRIPTION
**Pull request by mcfedr**
## Changes

changes the terraform module and provider loading to try to use use the versions endpoint as per the spec, if the extended api isn't available.

- https://www.terraform.io/internals/module-registry-protocol
- https://www.terraform.io/internals/provider-registry-protocol

The current code uses extensions to the spec, that are supported by registry.terraform.io, but not all third parties, e.g. GitLab. This will allow loading version info for modules hosted on GitLab and other registries that support the basic spec.

Also new endpoint doesn't have release date information, so thats now omitted when using these endpoints.

## Context

Fixes #14602

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
please trim sample JSON fixture to have less versions.

**Comment by viceice** [comment]
> @viceice I've reduced the size of the samples - just so you know, still changes the snapshots as the versions are in a random order in that response, so i've just left the first couple of 100 lines.

that's fine

File: lib/modules/datasource/terraform-provider/types.ts [hunk 1]
```diff
@@ -13,14 +13,10 @@ export interface TerraformBuild {
...
-export interface TerraformProvider {
-  namespace: string;
-  name: string;
-  provider: string;
-  source?: string;
-  versions: string[];
-  version: string;
-  published_at: string;
+export interface TerraformProviderVersions {
+  versions: Array<{
```
Comment (viceice) [inline lib/modules/datasource/terraform-provider/types.ts:17]: See other comment

File: lib/modules/datasource/terraform-provider/index.ts [hunk 2]
```diff
@@ -79,24 +79,14 @@ export class TerraformProviderDatasource extends TerraformDatasource {
...
-      dep.sourceUrl = res.source;
-    }
-    // set published date for latest release
-    const latestVersion = dep.releases.find(
-      (release) => res.version === release.version
-    );
-    // istanbul ignore else
-    if (latestVersion) {
-      latestVersion.releaseTimestamp = res.published_at;
-    }
```
Comment (viceice) [inline lib/modules/datasource/terraform-provider/index.ts:99]: Isn't this information available at some place?

File: lib/modules/datasource/terraform-module/index.ts [hunk 3]
```diff
@@ -63,24 +62,16 @@ export class TerraformModuleDatasource extends TerraformDatasource {
...
     const dep: ReleaseResult = {
-      releases: res.versions.map((version) => ({
+      releases: res.modules[0].versions.map(({ version }) => ({
         version,
       })),
     };
-    if (res.source) {
-      dep.sourceUrl = res.source;
+    if (res.modules[0].source) {
+      dep.sourceUrl = res.modules[0].source;
```
[the remark this answers is not in the document]
Comment (mcfedr) [reply lib/modules/datasource/terraform-module/index.ts:70]: So I noticed that it is in the GitLab implementation, but not in the API doc, or the registry.terraform.io version - which is where i took the snapshots from.

It was hard to tell what this information is used for later, maybe better to completely remove if its not too important?

File: lib/modules/datasource/terraform-module/index.ts [hunk 4]
```diff
@@ -63,24 +62,16 @@ export class TerraformModuleDatasource extends TerraformDatasource {
...
     if (pkgUrl.startsWith('https://registry.terraform.io/')) {
       dep.homepage = `https://registry.terraform.io/modules/${repository}`;
     }
-    // set published date for latest release
-    const latestVersion = dep.releases.find(
-      (release) => res.version === release.version
-    );
-    if (latestVersion) {
-      latestVersion.releaseTimestamp = res.published_at;
-    }
```
[the remark this answers is not in the document]
Comment (mcfedr) [reply lib/modules/datasource/terraform-module/index.ts:82]: yea, these APIs dont provide this information at all.

File: lib/modules/datasource/terraform-provider/index.ts (hunk 2 above)
Comment (mcfedr) [reply lib/modules/datasource/terraform-provider/index.ts:99]: same as above, these APIs are minimal, e.g. https://www.terraform.io/internals/provider-registry-protocol#sample-response

File: lib/modules/datasource/terraform-module/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/datasource/terraform-module/index.ts:70]: i wouldn't ask if it's not important. so we probably should have two implementations. one for official registry with old API but more metadata and the new for other registries. 🤔

i also use stability days, so if we loose created timestamps it would be worsening.

@rarkins WDYT?

[the remark this answers is not in the document]
Comment (mcfedr) [reply lib/modules/datasource/terraform-module/index.ts:70]: Yea, maybe that makes sense, I can leave the old implementation, and if 404, try the new version. Ill add this back in today.

**Comment by mcfedr** [comment]
@viceice so I've made a new commit - new try - this probably makes sense - tries the current endpoint first and if it doesnt work tries the documented api.

**Comment by viceice** [comment]
👉  https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

File: lib/modules/datasource/terraform-module/index.spec.ts [hunk 5]
```diff
@@ -79,6 +86,23 @@ describe('modules/datasource/terraform-module/index', () => {
...
+        .reply(404, {})
+        .get('/v1/modules/hashicorp/consul/aws/versions')
+        .reply(200, consulVersionsData)
+        .get('/.well-known/terraform.json')
+        .reply(200, serviceDiscoveryResult);
+      const res = await getPkgReleases({
+        datasource,
+        depName: 'hashicorp/consul/aws',
+      });
+      expect(res).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.spec.ts:102]: please use `toEqual` instead

File: lib/modules/datasource/terraform-module/index.spec.ts [hunk 6]
```diff
@@ -79,6 +86,23 @@ describe('modules/datasource/terraform-module/index', () => {
...
+      expect(res).not.toBeNull();
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.spec.ts:103]:
```suggestion
```

File: lib/modules/datasource/terraform-provider/index.spec.ts [hunk 7]
```diff
@@ -68,7 +75,24 @@ describe('modules/datasource/terraform-provider/index', () => {
...
-        .reply(200, JSON.parse(consulData))
+        .reply(200, JSON.parse(azurermData))
+        .get('/.well-known/terraform.json')
+        .reply(200, serviceDiscoveryResult);
+      const res = await getPkgReleases({
+        datasource: TerraformProviderDatasource.id,
+        depName: 'azurerm',
+      });
+      expect(res).toMatchSnapshot();
+      expect(res).not.toBeNull();
```
Comment (viceice) [inline lib/modules/datasource/terraform-provider/index.spec.ts:86]:
```suggestion
      expect(res).toMatchSnapshot();
```

use `toMatchObject` and something like `{ registryUrl: '.., ', ..., releases: expect.toBeArrayOfObjects().toBeArrayOfSize(30) }`

File: lib/modules/datasource/terraform-provider/index.ts [hunk 8]
```diff
@@ -79,6 +81,30 @@ export class TerraformProviderDatasource extends TerraformDatasource {
...
       registryURL
     );
+    try {
+      return await this.queryRegistryExtendedApi(
+        serviceDiscovery,
+        registryURL,
+        repository
+      );
+    } catch (err) {
+      return await this.queryRegistryVersions(
```
Comment (viceice) [inline lib/modules/datasource/terraform-provider/index.ts:91]: check if it's a 404 error and only try on those.
we should also not try this on official `registry.terraform.io`

**Review by viceice — COMMENTED** [review]
https://www.terraform.io/registry/api-docs#sample-request-5

**Review by viceice — CHANGES_REQUESTED** [review]
see comments

File: lib/modules/datasource/terraform-provider/index.ts (hunk 8 above)
Comment (mcfedr) [reply lib/modules/datasource/terraform-provider/index.ts:91]: I wrote this at first - but it seemed to be, if this url isnt part of the spec, a third party repo could easily reply with a 301 -> to the homepage or something similarly unhelpful, also this way, if the response from that that url is malformed, and there is a TypeError, it will still be able to get the information from the other url - for the provider url in particular, the url used in the original, `queryRegistryExtendedApi`, is not documented anywhere.

File: lib/modules/datasource/terraform-provider/index.spec.ts (hunk 7 above)
Comment (mcfedr) [reply lib/modules/datasource/terraform-provider/index.spec.ts:86]: is there a reason that snapshot testing has fallen out of favour? this is tests that I've not changed.

File: lib/modules/datasource/terraform-provider/index.ts (hunk 8 above)
Comment (mcfedr) [reply lib/modules/datasource/terraform-provider/index.ts:91]: The same applies to  terraform-module  - although its documented for registry.terraform.io its not for others, and there have clearly been problems with it, I assume thats why there is a sanity check on line 99

Comment (viceice) [reply lib/modules/datasource/terraform-provider/index.ts:91]: let's only use the old rich API requests for official registry, as there we know it's a official documented API and use the other for any other registry.

so no fallback at all and no duplicated calls

**Review by viceice — COMMENTED** [review]
see comment

File: lib/modules/datasource/terraform-provider/index.ts (hunk 8 above)
Comment (mcfedr) [reply lib/modules/datasource/terraform-provider/index.ts:91]: I was only able to find documentation of the rich api for modules, not for providers - but as its been working we could keep using it.

but yes, I see sense in skipping rich for 3rd parties, and just using offical.

so then its no longer a fall back, its just a 1st party, use old, 3rd party, use new. that does seem better for number of requests.

**Comment by mcfedr** [comment]
Any update on this? I think all the comments have been addressed @viceice

File: lib/modules/datasource/terraform-module/index.spec.ts [hunk 9]
```diff
@@ -4,6 +4,7 @@ import { loadFixture } from '../../../../test/util';
 import { TerraformModuleDatasource } from '.';
 
 const consulData: any = loadFixture('registry-consul.json');
+const consulVersionsData: any = loadFixture('registry-consul-versions.json');
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.spec.ts:7]:
```suggestion
const consulVersionsData = loadFixture('registry-consul-versions.json');
```
no any required

File: lib/modules/datasource/terraform-module/index.ts [hunk 10]
```diff
@@ -43,19 +49,56 @@ export class TerraformModuleDatasource extends TerraformDatasource {
       'terraform-module.getReleases()'
     );
 
+    const dep = await this.queryRegistry(registryUrl, registry, repository);
+
+    logger.trace({ dep }, 'dep');
+    return dep;
+  }
+
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.ts:57]: can you back inline this function? i think we don't need that extra function.

File: lib/modules/datasource/terraform-module/index.ts [hunk 11]
```diff
@@ -43,19 +49,56 @@ export class TerraformModuleDatasource extends TerraformDatasource {
...
-      const serviceDiscovery = await this.getTerraformServiceDiscoveryResult(
-        registryUrl
-      );
       pkgUrl = `${registry}${serviceDiscovery['modules.v1']}${repository}`;
       res = (await this.http.getJson<TerraformRelease>(pkgUrl)).body;
       const returnedName = res.namespace + '/' + res.name + '/' + res.provider;
       if (returnedName !== repository) {
         logger.warn({ pkgUrl }, 'Terraform registry result mismatch');
-        return null;
+        throw new Error('Terraform registry result mismatch');
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.ts:101]: why do we need to throw now?

File: lib/modules/datasource/terraform-module/index.ts [hunk 12]
```diff
@@ -80,8 +123,43 @@ export class TerraformModuleDatasource extends TerraformDatasource {
...
+    repository: string
+  ): Promise<ReleaseResult> {
+    let res: TerraformModuleVersions;
+    let pkgUrl: string;
+    try {
+      pkgUrl = `${registry}${serviceDiscovery['modules.v1']}${repository}/versions`;
+      res = (await this.http.getJson<TerraformModuleVersions>(pkgUrl)).body;
+      if (res.modules.length < 1) {
+        logger.warn({ pkgUrl }, 'Terraform registry result mismatch');
+        throw new Error('Terraform registry result mismatch');
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.ts:145]: i think we should return null instead of throwing a error which is caught right after.

File: lib/modules/datasource/terraform-module/index.ts [hunk 13]
```diff
@@ -80,8 +123,43 @@ export class TerraformModuleDatasource extends TerraformDatasource {
...
+
+    // Simplify response before caching and returning
+    const dep: ReleaseResult = {
+      releases: res.modules[0].versions.map(({ version }) => ({
+        version,
+      })),
+    };
+    if (res.modules[0].source) {
+      dep.sourceUrl = res.modules[0].source;
+    }
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.ts:159]: i think this is always undefined, so can be removed

File: lib/modules/datasource/terraform-module/index.ts [hunk 14]
```diff
@@ -80,8 +123,43 @@ export class TerraformModuleDatasource extends TerraformDatasource {
...
+    if (pkgUrl.startsWith('https://registry.terraform.io/')) {
+      dep.homepage = `https://registry.terraform.io/modules/${repository}`;
+    }
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.ts:162]: remove, should never be true and is wrong anyways

File: lib/modules/datasource/terraform-module/index.ts [hunk 15]
```diff
@@ -43,13 +49,39 @@ export class TerraformModuleDatasource extends TerraformDatasource {
       'terraform-module.getReleases()'
     );
 
+    const serviceDiscovery = await this.getTerraformServiceDiscoveryResult(
+      registryUrl
+    );
+    const registryHost = parseUrl(registryUrl)?.host;
+    if (registryHost === 'registry.terraform.io') {
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.ts:56]:
```suggestion
    if (registryUrl === defaultRegistryUrls[0]) {
```

Datasource layer will trim any trailing slashes

File: lib/modules/datasource/terraform-provider/index.spec.ts [hunk 16]
```diff
@@ -3,10 +3,11 @@ import * as httpMock from '../../../../test/http-mock';
...
 
-const consulData: any = loadFixture('azurerm-provider.json');
-const hashicorpReleases: any = loadFixture('releaseBackendIndex.json');
-const serviceDiscoveryResult: any = loadFixture('service-discovery.json');
-const telmateProxmocVersions: any = loadFixture(
+const azurermData = loadFixture('azurerm-provider.json');
+const azurermVersionsData = loadFixture('azurerm-provider-versions.json');
+const hashicorpReleases = loadFixture('releaseBackendIndex.json');
+const serviceDiscoveryResult = loadFixture('service-discovery.json');
+const telmateProxmocVersions = loadFixture(
```
Comment (viceice) [inline lib/modules/datasource/terraform-provider/index.spec.ts:10]: See other comment

File: lib/modules/datasource/terraform-module/index.ts [hunk 17]
```diff
@@ -36,21 +41,46 @@ export class TerraformModuleDatasource extends TerraformDatasource {
       return null;
     }
 
-    const { registry, repository } =
+    const { registry: registryUrlNormalized, repository } =
```
Comment (viceice) [inline lib/modules/datasource/terraform-module/index.ts:44]: why rename of variable?

File: lib/modules/datasource/terraform-provider/index.ts [hunk 18]
```diff
@@ -102,6 +122,28 @@ export class TerraformProviderDatasource extends TerraformDatasource {
...
+    const backendURL = `${registryURL}${serviceDiscovery['providers.v1']}${repository}/versions`;
+    const res = (await this.http.getJson<TerraformProviderVersions>(backendURL))
+      .body;
+    const dep: ReleaseResult = {
+      releases: res.versions.map(({ version }) => ({
+        version,
+      })),
+    };
+    dep.homepage = `${registryURL}/providers/${repository}`;
+    logger.trace({ dep }, 'dep');
```
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/datasource/terraform-provider/index.ts:137]: ☝️

File: lib/modules/datasource/terraform-module/index.ts (hunk 17 above)
Comment (mcfedr) [reply lib/modules/datasource/terraform-module/index.ts:44]: because when reviewing the code, its confusing that we have both registry and registryUrl and they are actually both urls,

File: lib/modules/datasource/terraform-provider/index.ts [hunk 19]
```diff
@@ -97,18 +113,37 @@ export class TerraformProviderDatasource extends TerraformDatasource {
...
+      })),
+    };
     return dep;
   }
 
   // TODO: add long term cache (#9590)
   private async queryReleaseBackend(
     packageName: string,
-    registryURL: string
+    registryUrl: string
```
Comment (viceice) [inline lib/modules/datasource/terraform-provider/index.ts:143]: please do those refactorings in separate PR's for easier review

Comment (mcfedr) [reply lib/modules/datasource/terraform-provider/index.ts]: fine. removed it. I just figured so much has changed anyway in this file

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Comment by mcfedr** [comment]
@viceice i guess something is wrong that it doesnt auto merge

**Comment by viceice** [comment]
nope, needs update from main. 😉
